task1 na review

patrikprovaznik opened the pull request.

patrikprovaznik committed:
task1 na review

mirokovac commented on 1task.py:
```diff
@@ -0,0 +1,60 @@
+import json
```
@patrikprovaznik rename this file to something more specific like api.py or interface_api.py

mirokovac commented on task-1-interfaces.json:
```diff
@@ -0,0 +1,231 @@
+{
```
@patrikprovaznik keep this in resource folder

mirokovac commented on task-1.txt:
```diff
@@ -0,0 +1,27 @@
+In this task preson should learne how to create webserver using Flask library (Optional use django if you are more familiar with that). It s pure procedural programing using well defined data un json format. These data
```
@patrikprovaznik please remove this from project. it is not needed in here.

mirokovac commented on 1task.py:
```diff
@@ -0,0 +1,60 @@
+app = Flask(__name__)
+
+
+@app.route('/get-all-interfaces', methods=['GET'])
```
@patrikprovaznik extra task - please provide me with extra method to reload default configuration where configuration default one is in json file.

This means you have to open the json file in the beginning and cache it for yourself. and use cached data to show add or delete.

mirokovac commented on 1task.py:
```diff
@@ -0,0 +1,60 @@
+
+
+@app.route('/get-interface/<path:interface_name>', methods=['GET'])
+def get_some_interface(interface_name):
```
@patrikprovaznik learn about typing module and typehints.. depends on project but we usually try and use them.. it s a good practice. please use it throughout all the tasks.

mirokovac commented on 1task.py:
```diff
@@ -0,0 +1,60 @@
+def get_all_interfaces():
+    with open('task-1-interfaces.json') as f:
+        data = json.loads(f.read())
+        interfaces = data['ietf-interfaces:interfaces']['interface']
```
@patrikprovaznik what if there are no data? use dict get method something like
```data,get('ietf-interfaces:interfaces', {}).get('interface', [])```
this way you make sure that you always get at least empty list

also when you ll be opening your json file once at the beginning to cache it and every time for new default method dont forget to try except if the file doesnt exist.

mirokovac commented on 1task.py:
```diff
@@ -0,0 +1,60 @@
+        interfaces = data['ietf-interfaces:interfaces']['interface']
+        for interface in interfaces:
+            if interface["name"] == interface_name:
+                return jsonify(interface)
```
@patrikprovaznik do you think we can cache these by the name separately as well to receive single interface much faster? if you do don t forget to remove it from cache also when deleting

mirokovac commented on 1task.py:
```diff
@@ -0,0 +1,60 @@
+        for interface in interfaces:
+            if interface["name"] == interface_name:
+                return jsonify(interface)
+    abort(404)
```
@patrikprovaznik find ```from http import HTTPStatus``` library which has these codes already implemented

mirokovac commented on 1task.py:
```diff
@@ -0,0 +1,60 @@
+
+@app.route('/delete-interface/<path:interface_name>', methods=['DELETE'])
+def delete_interface(interface_name):
+    with open('task-1-interfaces.json') as f:
+        data = json.loads(f.read())
+        interfaces = data['ietf-interfaces:interfaces']['interface']
+        for interface in interfaces:
+            if interface["name"] == interface_name:
+                interfaces.remove(interface)
+                return jsonify("DELETE was Successful.")
+    return jsonify("DELETE was Unsuccessful")
```
@patrikprovaznik if you would do it this way get would always give us all the interfaces even ones that you have removed. you would have to remove them from json file.. now we decided to cache it please don t forget to remove it from caches

mirokovac replied:
@patrikprovaznik return 404 not found if the interface doesn t exist.. ([for, else clause could help you here???](https://book.pythontips.com/en/latest/for_-_else.html))

mirokovac reviewed (changes requested):
@patrikprovaznik put .idea file into .gitignore file so we don t have this in repository.. this is keeping structure of your local IDE which everyone might have different and we don t want to share that with other people. more about gitignore you can find in [here](https://git-scm.com/docs/gitignore)

Also I mentioned caching. when you start working on them please override Flask class and inside of init make this work. something like ->

```
InterfaceAPI(Flask):
    def __init__(self, import_name):
        super(InterfaceAPI, self).__init__(import_name)
        load_your_cache_in_here
```

One more suggestion - in majority of projects when you are writing commit message it should be descriptive in what has changed. you can find best practice in [here](https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53). first line is shor summary and then you type what you did but don t go into a specific details or anything just some short description one to three sentences more or less.

patrikprovaznik committed:
New important changes have been made to the code for example:
1. adding a class in which we created the init method, which is used to load the json file into the cache memory at the beginning of the program.
2. after creating the cache memory we had to adjust all created functions so that they take data from this cache memory.
3. next we made changes such as: removing .txt file (assignment of task1), putting .idea file into .gitignore, adding type hints, etc.

mirokovac commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
+.idea/
 # Byte-compiled / optimized / DLL files
```
@patrikprovaznik this file needs to be in the root directory in order for .idea folder to be ignored..

mirokovac commented on interface_api.py:
```diff
@@ -0,0 +1,64 @@
+
+
+@app.route('/get-interface/<path:interface_name>', methods=['GET'])
+def get_some_interface(interface_name: str) -> Response | tuple[str, int]:
```
@patrikprovaznik I don t think it s a problem I just haven t seen it used this way yet.. We usually use Union
```
Union[Response, Tuple[str, int]]
```
anyway for sure don t use tuple but Tuple with capital T imported from typing module

mirokovac commented on interface_api.py:
```diff
@@ -0,0 +1,64 @@
+    all_interfaces = app.cache.values()
+    req_data = request.get_json()
+    result = []
+    for interface in all_interfaces:
+        for input_interface in req_data['input']['interfaces']:
+            interface_name = input_interface.get("name")
+            interface_type = input_interface.get("type")
+            interface_enabled = input_interface.get("enabled")
+            if (interface_name is None or interface["name"] == interface_name) and \
+                    (interface_type is None or interface["type"] == interface_type) and \
+                    (interface_enabled is None or interface["enabled"] == interface_enabled):
+                if interface not in result:
+                    result.append(interface)
```
@patrikprovaznik this can be pretty performance heavy but you have to remember that there is almost always something you do when you cycle many times over some things (99.9 percent of time) and usually if you can t then the design of code is not right and should think of redesigning.

my first rule is to never use more then two cycles in each other. you have two which is ok. if you have more it s usually good idea the longest one change to dictionary and try to get data somehow from that instead of looping all the time through all of them.

secondly always looks for some conditions that can break your for loop or continue if you can somehow so only part of cycle will be actually computed and rest will be skipped.

So when you cycling over your interfaces at the end you check if `interface not in result`. which means if you add it in you can break cycle and continue to another interface cause even if it will comply with another condition of your input it will not be added second time anymore. which means that your `if interface not in result` will be useless as well if you break afterwards. same interface will not be in there anymore. This will lower the amount of cycling immensely, especially if first condition will be already true and there is like 10 more options 


It looks like redundant thing in here cause you don t have much of interfaces but usually in projects we have lists of 10 000 things and have similar problem to this

mirokovac commented on interface_api.py:
```diff
@@ -0,0 +1,64 @@
+            interface_name = input_interface.get("name")
+            interface_type = input_interface.get("type")
+            interface_enabled = input_interface.get("enabled")
+            if (interface_name is None or interface["name"] == interface_name) and \
+                    (interface_type is None or interface["type"] == interface_type) and \
+                    (interface_enabled is None or interface["enabled"] == interface_enabled):
```
@patrikprovaznik 
```
interface["name"], interface["type"], interface["enabled"]
```
use rather get. especially when you are not sure if the can exist always.. in json you saw that they can but if different device they can be different. with interfaces usually only name is mandatory

mirokovac commented on interface_api.py:
```diff
@@ -0,0 +1,64 @@
+def delete_interface(interface_name: str) -> tuple[str, int]:
+    interface_deleted = app.cache.pop(interface_name, None)
+    if interface_deleted:
+        return 'DELETE was Successful', http.HTTPStatus.OK
```
@patrikprovaznik minor thing but it would be nice to have in a message the name of the interface that got deleted even if it is the path. if anything goes wrong (some wrong interface would get deleted etc... ) you would see that in the message that you asked for different interface to delete then you actually deleted..

mirokovac reviewed (changes requested):
@patrikprovaznik Much better. my biggest problem in here would be the cycling  which you can easily fix in here. but nice work!

patrikprovaznik committed:
Some final changes have been made to the code for example:
1. adding break to nested loops in post_several_interfaces function to lower the amount of cycling
2. removing condition - if not in result because adding break in this function solve this problem too + we also added get() for input_interface
3. adding message with name of deleted/not deleted interface in delete_interface function
4. next we made small changes such as: putting .gitignore into root director and adding recommended type hints

patrikprovaznik committed:
moved to separated folder

mirokovac commented on .idea/.gitignore:
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
```
ignore .idea folder

patrikprovaznik committed:
adding .idea/ and *.log into gitignore

patrikprovaznik committed:
removed .idea

mirokovac commented on task1/interface_api.py:
```diff
@@ -0,0 +1,65 @@
+        with open('resource/task-1-interfaces.json') as f:
+            data = json.loads(f.read())
+            interfaces = data['ietf-interfaces:interfaces']['interface']
+            for interface in interfaces:
+                self.cache[interface['name']] = interface
```
@patrikprovaznik try dict comperhension

mirokovac commented on task1/interface_api.py:
```diff
@@ -0,0 +1,65 @@
+    if interface:
+        return jsonify(interface)
+    else:
+        return f'Interface {interface_name} does not exist.', http.HTTPStatus.NOT_FOUND
```
@patrikprovaznik return this as a json so it can be processed easily for client app if needed

mirokovac commented on task1/interface_api.py:
```diff
@@ -0,0 +1,65 @@
+    all_interfaces = app.cache.values()
+    req_data = request.get_json()
+    result = []
+    for interface in all_interfaces:
+        for input_interface in req_data['input']['interfaces']:
```
@patrikprovaznik invert this cycle and if we have name of input request then get it from your cache.. which is much faster if we don t need to iterate through all interfaces.

mirokovac commented on task1/interface_api.py:
```diff
@@ -0,0 +1,65 @@
+def delete_interface(interface_name: str) -> Union[Response, Tuple[str, int]]:
+    interface_deleted = app.cache.pop(interface_name, None)
+    if interface_deleted:
+        return f'DELETE WAS SUCCESSFUL.\nThe name of the deleted interface is: {interface_name}', http.HTTPStatus.OK
```
@patrikprovaznik again some json like {info: msg}

mirokovac commented on task1/interface_api.py:
```diff
@@ -0,0 +1,65 @@
+    if interface_deleted:
+        return f'DELETE WAS SUCCESSFUL.\nThe name of the deleted interface is: {interface_name}', http.HTTPStatus.OK
+    else:
+        return f'DELETE WAS UNSUCCESSFUL.\nInterface {interface_name} does not exist.', http.HTTPStatus.NOT_FOUND
```
@patrikprovaznik same as above

mirokovac reviewed (changes requested):
@patrikprovaznik small performance fixes needed

patrikprovaznik committed:
a couple of changes have been made:
1.  returning output of functions as a json, so it can be now processed for client app if needed
2.  we inverted cycle in function - post_several_interfaces. So now if we have name of input request we are getting it from cache, and we don't need to iterate through all interfaces.

patrikprovaznik committed:
one change has been made:
 static method for conditions from InterfaceAPI class was removed, and we added these conditions directly into post_several_interfaces function
